### PR TITLE
Makefile, actions workflows: use cargo sparse index protocol

### LINKED ISSUE #3147: Use the `sparse` cargo indexing protocol whereever we can
**What I'd like:**
To use the `sparse` cargo indexing protocol whereever we can in Bottlerocket builds, including Github actions workflows, `cargo-make` tasks, etc.

The time savings for freshly indexing `crates.io` is pretty significant. See https://blog.rust-lang.org/inside-rust/2023/01/30/cargo-sparse-protocol.html

**Any alternatives you've considered:**
N/A

### PR DESCRIPTION
**Pull request by etungsten**
**Issue number:**

Resolves https://github.com/bottlerocket-os/bottlerocket/issues/3147

**Description of changes:**

```
   Makefile, actions workflows: use cargo sparse index protocol
    
    Switch to using cargo's sparse indexing protocol for accessing
    crates.io. This gives a significant performance boost when cargo first
    retrieves the index from crates.io.
```

**Testing done:**

See the checks for this PR.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -29,6 +29,9 @@ env:
   # This is a useful early signal during GitHub actions to see if there are
   # upstream Go code problems.
   GOPROXY: direct
+  # Enables cargo's sparse indexing protocol which significantly boots indexing performance
```
Comment (zmrow) [inline .github/workflows/build.yml]: s/boots/boosts

File: .github/workflows/cache.yml [hunk 2]
```diff
@@ -13,6 +13,10 @@ on:
       - '!tools/pubsys/**.example'
       - '!tools/pubsys/**.template'
       - 'tools/Cargo.lock'
+env:
+  # Enables cargo's sparse indexing protocol which significantly boots indexing performance
```
Comment (zmrow) [inline .github/workflows/cache.yml]: Same here: s/boots/boosts

File: Makefile.toml [hunk 3]
```diff
@@ -96,6 +96,9 @@ BUILDSYS_UPSTREAM_LICENSE_FETCH= "false"
 BUILDSYS_JOBS = "8"
 
 CARGO_HOME = "${BUILDSYS_ROOT_DIR}/.cargo"
+# Enables cargo's sparse indexing protocol which significantly boots indexing performance
```
Comment (zmrow) [inline Makefile.toml]: One more

**Comment by etungsten** [comment]
Taking into draft since CI is failing over something I can't repro locally.

**Comment by etungsten** [comment]
Closing this in favor of switching to [rust 1.70.0](https://blog.rust-lang.org/2023/06/01/Rust-1.70.0.html#sparse-by-default-for-cratesio) where the indexing protocol is set to `sparse` by default.

**Pull request closed without merging**